src/core/sock/sockinfo_tcp.cpp

```diff
@@ -6061,13 +6234,63 @@ int sockinfo_tcp::tcp_tx_express(const struct iovec *iov, unsigned iov_len, uint
     }
     if (!(flags & XLIO_EXPRESS_MSG_MORE)) {
         tcp_output(&m_pcb);
+        m_b_xlio_socket_dirty = false;
+    } else if (m_p_group && !m_b_xlio_socket_dirty) {
+        m_b_xlio_socket_dirty = true;
+        m_p_group->add_dirty_socket(this);
     }
 
     unlock_tcp_con();
 
     return bytes_written;
 }
 
+int sockinfo_tcp::tcp_tx_express_inline(const struct iovec *iov, unsigned iov_len, unsigned flags)
+{
+    if (unlikely(!is_connected_and_ready_to_send())) {
+        return -1;
+    }
+
+    pbuf_desc mdesc;
+    int bytes_written = 0;
+
+    memset(&mdesc, 0, sizeof(mdesc));
+    mdesc.attr = PBUF_DESC_NONE;
+
+    lock_tcp_con();
+
+    for (unsigned i = 0; i < iov_len; ++i) {
+        bytes_written += iov[i].iov_len;
+        err_t err = tcp_write(&m_pcb, iov[i].iov_base, iov[i].iov_len, 0, &mdesc);
+        if (unlikely(err != ERR_OK)) {
+            // XXX tcp_write() can return multiple errors.
+            // XXX tcp_write() can also fail due to queuelen limit, but this is unlikely.
+            m_conn_state = TCP_CONN_ERROR;
+            m_error_status = ENOMEM;
+            return tcp_tx_handle_errno_and_unlock(ENOMEM);
+        }
+    }
+    if (!(flags & XLIO_EXPRESS_MSG_MORE)) {
+        m_b_xlio_socket_dirty = false;
+        tcp_output(&m_pcb);
+    } else if (m_p_group && !m_b_xlio_socket_dirty) {
```

Comment:
Should we check here if dirty handling is requested?

Reply:
Currently it's checked inside the poll_group method, because this is the group configuration whether to track dirty sockets or not.